### ISSUE TITLE
Cupertino pickers are not focusable with a keyboard

### ISSUE THREAD

**Issue by gmoraja** [flutter/flutter#49811]
@goderbauer
@gspencergoog
@lorenzOliveto

**Comment by Hixie** [comment]
@darrenaustin is this something you can look into?
@gspencergoog does any of your recent work make this trivial?

**Comment by gspencergoog** [comment]
Well, I don't know about _trivial_, but [`FocusableActionDetector`](https://api.flutter.dev/flutter/widgets/FocusableActionDetector-class.html) makes it pretty compact to implement focus, hover, shortcuts, and activation actions.

**Comment by xster** [comment]
More issues got in the way. No progress yet. Still in my queue.

**Comment by xster** [comment]
I wasn't very familiar with our current keyboard / focus work so I brushed up a bit. 

However, I wound up in a rabbit hole a bit. The "correct" solution is non-obvious and likely multi-quarter so we'll likely need some more guidance on the specific blocker @gmoraja.

To make a long story long, 

1. What the right focus highlight UI should be is platform dependent. What the focus indicator looks like on macOS ([FR](https://github.com/flutter/flutter/issues/43365) not yet implemented) is different from what the focus indicator looks like on iPadOS ([FR](https://github.com/flutter/flutter/issues/53456) not yet implemented), and is different from what the focus indicator looks like on Android and web (I assume we made a lot more progress there @gspencergoog). @gmoraja what circumstance is your primary use case? We can focus on unblocking on that platform first than having all options for all platforms. 
2. Specifically for CupertinoPicker and the upstream UIPickerView, it's also not clear what the focus highlight UI should be. UIPickerView itself is not keyboard focusable AFAICT. You can't focus onto it and operate with a keyboard on iPadOS and there are no UIPickerViews in AppKit for macOS. The accessibility highlight for VoiceOver for iOS is not very pretty so we probably don't want to follow that example for keyboard focus

![flutter_01](https://user-images.githubusercontent.com/156888/77811537-382b0d80-7058-11ea-923e-432e30f4643d.png)

I think we'll have to create something custom not based on Apple spec for this unfortunately.

**Comment by gmoraja** [comment]
@xster we’re using CupertinoPickers and CupertinoDatePickers on Android to let users choose from a list of elements (CupertinoPickers) or a duration in hours and minutes (CupertinoDatePickers).
Our app needs to be fully accessible so all the sections need to be navigable with a keyboard.

**Comment by xster** [comment]
I meant that it's unclear how the focus highlight should visually appear since there are no prior art. But we could just make up something non-standard (since there are no standards). WDYT?

**Comment by gspencergoog** [comment]
If nothing else, just implementing something non-standard would give us a place to plug in a "correct" solution in the future, and provide some functionality in the meantime.

**Comment by TahaTesser** [comment]
Using the  Apple magic keyboard on iOS 14, native picker focused but not in Flutter

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/cupertino.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      title: 'Cupertino App',
      home: CupertinoPageScaffold(
        navigationBar: CupertinoNavigationBar(
          middle: Text('Cupertino App Bar'),
        ),
        child: Center(
          child: Container(
            height: 300,
            child: CupertinoDatePicker(
                mode: CupertinoDatePickerMode.dateAndTime,
                onDateTimeChanged: (dateTime) {}),
          ),
        ),
      ),
    );
  }
}

```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.22.0-10.0.pre.204, on Mac OS X 10.15.6 19G2021,
    locale en-GB)
    • Flutter version 1.22.0-10.0.pre.204 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision b2b5f2636a (23 hours ago), 2020-09-16 08:02:06 -0700
    • Engine revision 1ef10b240e
    • Dart version 2.10.0 (build 2.10.0-124.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.0
    • macOS (desktop)        • macos                     • darwin-x64     • Mac
      OS X 10.15.6 19G2021
    • Web Server (web)       • web-server                • web-javascript •
      Flutter Tools
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 85.0.4183.102

• No issues found!

```

</details>

**Comment by hkwinslow** [comment]
Could someone please answer for me whether Flutter is supposed to support navigation with a bluetooth keyboard in iOS? I have read through many Github issues and seem to get conflicting answers. I cannot access my Flutter app via a keyboard. I would like to tab to different input elements but cannot seem to figure out how to do this. Thank you.

**Comment by maheshmnj** [comment]
Reproducible on stable 3.3 and master 3.7

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel master, 3.7.0-13.0.pre.131, on macOS 13.1 22C65 darwin-arm64, locale en-IN)
    • Flutter version 3.7.0-13.0.pre.131 on channel master at /Users/mahesh/Development/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0196e6050b (2 days ago), 2022-12-31 11:10:23 -0500
    • Engine revision 932591ec04
    • Dart version 3.0.0 (build 3.0.0-76.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks
      and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
    • All required HTTP hosts are available
    
• No issues found!
```
```
[✓] Flutter (Channel stable, 3.3.9, on macOS 12.6 21G115 darwin-arm, locale en-IN)
    • Flutter version 3.3.9 on channel stable at /Users/mahesh/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (24 hours ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.53.20221101

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>
